### PR TITLE
[content-visibility] Lazily layout children

### PR DESCRIPTION
**Pull request by rwlbuis**
#### faa0935aebd5fb9dfde94bf8c91262fd70479bc4
<pre>
[content-visibility] Lazily layout children
<a href="https://bugs.webkit.org/show_bug.cgi?id=237369">https://bugs.webkit.org/show_bug.cgi?id=237369</a>

Reviewed by Tim Nguyen.

This patch supports skipping layout of children in hidden content-visibility subtrees
for normal layout, since they do not count as content [1].

However DOM operations that require measuring do need an accurate layout, so
introduce the helper class ContentVisibilityForceLayoutScope to override (&quot;forced layout&quot;)
the skipped layout mechanism for hidden content-visibility subtrees.

Roughly the skipping layout of children falls into two catagories:
- general block flows (note flexbox is part of this)
- grid containers

The existing perf tests under ContentVisibility/ improve a lot with this patch (up to 100x), but
the scrolling variants have only slight improvement since scrolling content into the viewport
means falling back to full layout.
Improving the perf tests of grid based content-visibility perf tests will be addressed in a future PR.

[1] <a href="https://w3c.github.io/csswg-drafts/css-contain/#skips-its-contents">https://w3c.github.io/csswg-drafts/css-contain/#skips-its-contents</a>

* LayoutTests/TestExpectations:
* Source/WebCore/css/ComputedStyleExtractor.cpp: make sure layout is up-to-date for hidden content-visibility subtree.
(WebCore::ComputedStyleExtractor::propertyValue const):
* Source/WebCore/dom/Document.cpp:
(WebCore::Document::updateLayout): forces layout on hidden content-visibility subtree if needed.
(WebCore::Document::updateLayoutIgnorePendingStylesheets):
(WebCore::Document::updateLayoutIfDimensionsOutOfDate):
* Source/WebCore/dom/Document.h:
* Source/WebCore/dom/Element.cpp:
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::offsetLeftForBindings):
(WebCore::Element::offsetTopForBindings):
(WebCore::Element::scrollLeft):
(WebCore::Element::scrollTop):
(WebCore::Element::setScrollLeft):
(WebCore::Element::setScrollTop):
(WebCore::Element::getClientRects):
(WebCore::Element::boundingClientRect):
(WebCore::Element::innerText):
* Source/WebCore/html/HTMLImageElement.cpp:
(WebCore::HTMLImageElement::width):
(WebCore::HTMLImageElement::height):
(WebCore::HTMLImageElement::x const):
(WebCore::HTMLImageElement::y const):
* Source/WebCore/html/ImageInputType.cpp:
(WebCore::ImageInputType::height const):
(WebCore::ImageInputType::width const):
* Source/WebCore/page/LocalFrameViewLayoutContext.h: keep track of forced layout.
(WebCore::LocalFrameViewLayoutContext::needsSkippedContentLayout const):
* Source/WebCore/rendering/RenderBlock.cpp:
(WebCore::RenderBlock::layoutPositionedObject): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderBlockFlow.cpp:
(WebCore::RenderBlockFlow::layoutInFlowChildren): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::clearNeedsLayoutForDescendants): add helper to clear layout for subtree.
(WebCore::RenderElement::layoutIfNeeded): take isSkippedContentForLayout into account.
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::layoutIfNeeded): Deleted.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutPositionedObject): skip layout in hidden content-visibility subtree.
* Source/WebCore/rendering/RenderLayoutState.cpp: Add helper to force layout if there is a hidden content-visibility subtree.
(WebCore::ContentVisibilityForceLayoutScope::ContentVisibilityForceLayoutScope):
(WebCore::ContentVisibilityForceLayoutScope::~ContentVisibilityForceLayoutScope):
* Source/WebCore/rendering/RenderLayoutState.h:
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::isSkippedContentForLayout const): same as isSkippedContent but takes into account needsSkippedContentLayout.
* Source/WebCore/rendering/RenderObject.h:
* Source/WebCore/svg/SVGGeometryElement.cpp:
(WebCore::SVGGeometryElement::getTotalLength const):
(WebCore::SVGGeometryElement::getPointAtLength const):
(WebCore::SVGGeometryElement::isPointInFill):
(WebCore::SVGGeometryElement::isPointInStroke):
* Source/WebCore/svg/SVGLocatable.cpp:
(WebCore::SVGLocatable::getBBox):
(WebCore::SVGLocatable::computeCTM):
* Source/WebCore/svg/SVGPathElement.cpp:
(WebCore::SVGPathElement::getBBox):
* Source/WebCore/svg/SVGSVGElement.cpp:
(WebCore::SVGSVGElement::getIntersectionList):
(WebCore::SVGSVGElement::getEnclosureList):
(WebCore::SVGSVGElement::checkIntersection):
(WebCore::SVGSVGElement::checkEnclosure):
* Source/WebCore/svg/SVGTextContentElement.cpp:
(WebCore::SVGTextContentElement::getNumberOfChars):
(WebCore::SVGTextContentElement::getComputedTextLength):
(WebCore::SVGTextContentElement::getCharNumAtPosition):
* Source/WebCore/testing/Internals.cpp:
(WebCore::Internals::updateLayoutIgnorePendingStylesheetsAndRunPostLayoutTasks):

Canonical link: <a href="https://commits.webkit.org/268310@main">https://commits.webkit.org/268310@main</a>
</pre>

https://github.com/WebKit/WebKit/commit/784bc3e674082d945826aa7ce4e62e55a882516e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/19307 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/19727 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/20324 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/21198 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/18065 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/19540 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/22996 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/19850 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/19708 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/19526 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/19577 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/16785 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/22066 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/16763 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/17573 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/23911 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/17825 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/17748 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/21877 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/18340 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/15528 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/17473 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4617 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/21833 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/18170 "Built successfully") | | |

### CODE REVIEW

**Comment by rwlbuis** [comment]
Sorry, the first time this was put up the patch was not very good, I forgot about skipping more in inline children layout and some other things, only now I feel the patch is ready.

File: Source/WebCore/page/FrameViewLayoutContext.h [hunk 1]
```diff
@@ -82,6 +82,8 @@ class FrameViewLayoutContext {
     bool isInLayout() const { return layoutPhase() != LayoutPhase::OutsideLayout; }
     bool isInRenderTreeLayout() const { return layoutPhase() == LayoutPhase::InRenderTreeLayout; }
     bool inPaintableState() const { return layoutPhase() != LayoutPhase::InRenderTreeLayout && layoutPhase() != LayoutPhase::InViewSizeAdjust && (layoutPhase() != LayoutPhase::InPostLayout || inAsynchronousTasks()); }
+    bool isLayoutForced() const { return m_layoutForced; }
```
Comment (smfr) [inline Source/WebCore/page/FrameViewLayoutContext.h]: "Force" here is too vague. Forced by what? This needs a name that clearly associates it with skipped content.

**Review by smfr — DISMISSED** [review]

File: Source/WebCore/page/FrameViewLayoutContext.h (hunk 1 above)
Comment (rwlbuis) [reply Source/WebCore/page/FrameViewLayoutContext.h]: Fair enough, I renamed to something more descriptional.

**Comment by alanbaradlay** [comment]
I am not sure how maintainable this approach is in the long run. I would surely make the mistake of not protecting a newly introduced updateLayout() call with this skipped scope. Unfortunately I don't have a better idea at this point either.

**Comment by rwlbuis** [comment]
> I am not sure how maintainable this approach is in the long run. I would surely make the mistake of not protecting a newly introduced updateLayout() call with this skipped scope. Unfortunately I don't have a better idea at this point either.

AFAICS this only affects JS exposed DOM methods though, these new methods should not be added that often. In general, would it help here to have WPT tests test a lot more methods (this patch touched the minimum amount of JS exposed methods to pass the current content-visibility WPT tests and keep the patch size as small as possible, but I don't think it would make the patch that big to add the methods I know need updating).

**Comment by smfr** [comment]
> > I am not sure how maintainable this approach is in the long run. I would surely make the mistake of not protecting a newly introduced updateLayout() call with this skipped scope. Unfortunately I don't have a better idea at this point either.
> 
> AFAICS this only affects JS exposed DOM methods though, these new methods should not be added that often.

But how do we know what these are, other than relying on (probably always incomplete) WPT coverage? The spec doesn't list them all (in fact, reading the spec I didn't find anything that said that JS geometry queries should require this behavior).

**Comment by alanbaradlay** [comment]
> > > I am not sure how maintainable this approach is in the long run. I would surely make the mistake of not protecting a newly introduced updateLayout() call with this skipped scope. Unfortunately I don't have a better idea at this point either.
> > 
> > 
> > AFAICS this only affects JS exposed DOM methods though, these new methods should not be added that often.
> 
> But how do we know what these are, other than relying on (probably always incomplete) WPT coverage? The spec doesn't list them all (in fact, reading the spec I didn't find anything that said that JS geometry queries should require this behavior).
It is mentioned at the general description of https://w3c.github.io/csswg-drafts/css-contain/#content-visibility
"and do not respond to hit-testing (as if they had), and to a large extent do not update their styles at all unless explicitly requested by script"
but I totally agree on the highly error prone part.

File: Source/WebCore/dom/Document.cpp [hunk 2]
```diff
@@ -2282,6 +2282,31 @@ bool Document::updateStyleIfNeeded()
...
+{
+    if (m_document && m_document->view())
+        m_document->view()->layoutContext().forceSkippedContentLayout(false);
+}
+
+void Document::updateLayoutForNode(const Element& element)
+{
+    ContentVisibilityForceLayoutScope scope(element);
+    updateLayoutIgnorePendingStylesheets();
+}
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]: Can we not introduce this new method and instead have a single entry layout entry point that takes an OptionSet ?

It's possible someone might want to update layout not ignoring pending stylesheets, while also taking in account content-visibility. Also, with a single point of entry with options, people will have to think about the `OptionSet<>` and whether they want content-visibility to be taken in account every time they add a new layout.

File: Source/WebCore/dom/Document.cpp [hunk 3]
```diff
@@ -2282,6 +2282,31 @@ bool Document::updateStyleIfNeeded()
...
 
+ContentVisibilityForceLayoutScope::ContentVisibilityForceLayoutScope(const Element& context)
+{
+    if (context.document().view()) {
+        context.document().updateRelevancyOfContentVisibilityElements();
+        context.document().updateStyleIfNeeded();
+        if (context.renderer() && context.renderer()->style().skippedContentReason().has_value()) {
+            context.renderer()->setNeedsLayout();
+            m_document = context.document();
+            m_document->view()->layoutContext().forceSkippedContentLayout(true);
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]: The name `forceSkippedContentLayout` is slightly confusing, we're not forcing a layout, but setting a flag.

Maybe `setNeedsSkippedContentLayout()` && `needsSkippedContentLayout()`

File: Source/WebCore/rendering/RenderBlockFlow.cpp [hunk 4]
```diff
@@ -746,12 +746,15 @@ void TextBoxTrimmer::adjustTextBoxTrimAfterLayout()
 
 void RenderBlockFlow::layoutInFlowChildren(bool relayoutChildren, LayoutUnit& repaintLogicalTop, LayoutUnit& repaintLogicalBottom, LayoutUnit& maxFloatLogicalBottom)
 {
-    if (!firstChild()) {
+    if (!firstChild() || (firstChild()->isSkippedContentForLayout() && !normalChildNeedsLayout())) {
```
Comment (nt1m) [inline Source/WebCore/rendering/RenderBlockFlow.cpp]: Same question here: can you elaborate why these changes are needed? Are the changes in `layoutIfNeeded` not good enough to skip work here?

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp
Comment (nt1m) [inline Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:1]: The changes in this file seem a bit arbitrary, can you elaborate why they are needed? Are the changes in `layoutIfNeeded` not good enough to skip work here?

File: Source/WebCore/page/LocalFrameViewLayoutContext.h [hunk 5]
```diff
@@ -93,6 +93,9 @@ class LocalFrameViewLayoutContext {
     bool isInRenderTreeLayout() const { return layoutPhase() == LayoutPhase::InRenderTreeLayout; }
     bool inPaintableState() const { return layoutPhase() != LayoutPhase::InRenderTreeLayout && layoutPhase() != LayoutPhase::InViewSizeAdjust && (layoutPhase() != LayoutPhase::InPostLayout || inAsynchronousTasks()); }
 
+    bool isSkippedContentLayoutForced() const { return m_isSkippedContentLayoutForced; }
+    void forceSkippedContentLayout(bool f) { m_isSkippedContentLayoutForced = f; }
+
```
Comment (nt1m) [inline Source/WebCore/page/LocalFrameViewLayoutContext.h]: These should probably be on `RenderLayoutState`

File: Source/WebCore/svg/SVGGeometryElement.cpp [hunk 6]
```diff
@@ -47,7 +47,7 @@ SVGGeometryElement::SVGGeometryElement(const QualifiedName& tagName, Document& d
 
 float SVGGeometryElement::getTotalLength() const
 {
-    document().updateLayoutIgnorePendingStylesheets();
+    document().updateLayoutForNode(*this);
```
Comment (nt1m) [inline Source/WebCore/svg/SVGGeometryElement.cpp]: Does content-visibility properly work on SVG in general?

File: Source/WebCore/dom/Element.cpp [hunk 7]
```diff
@@ -3794,7 +3799,7 @@ ExceptionOr<void> Element::setInnerHTML(const String& html)
 String Element::innerText()
 {
     // We need to update layout, since plainText uses line boxes in the render tree.
-    document().updateLayoutIgnorePendingStylesheets();
+    document().updateLayoutForNode(*this);
```
Comment (nt1m) [inline Source/WebCore/dom/Element.cpp]: Do we need this for innerText() ? Since skipped content is skipped from innerText, I'm not sure we need to lay it out (I may be wrong though)

File: Source/WebCore/dom/Element.cpp [hunk 8]
```diff
@@ -1318,6 +1318,8 @@ static HashSet<TreeScope*> collectAncestorTreeScopeAsHashSet(Node& node)
 
 int Element::offsetLeftForBindings()
```
Comment (nt1m) [inline Source/WebCore/dom/Element.cpp]: Is it correct to do this just for script? Do we know if some internal callers may want skipped content layout as well ?

**Review by nt1m — COMMENTED** [review]
Here's some high-level feedback after discussing with other folks:
- We'd like to make sure future code also takes content-visibility in account, potentially through a common layout entry point that takes an OptionSet that needs to be thought out.

- Do we know of places that are not script that might want skipped content to be laid out? (find in page for instance)

File: Source/WebCore/html/ImageInputType.cpp [hunk 9]
```diff
@@ -175,6 +175,8 @@ unsigned ImageInputType::height() const
     ASSERT(element());
     Ref<HTMLInputElement> element(*this->element());
 
+    ContentVisibilityForceLayoutScope scope(element.get());
```
Comment (smfr) [inline Source/WebCore/html/ImageInputType.cpp]: These seems to be sprinkled around fairly arbitrarily. How will future patches know when this is needed? We need to factor the code in such a way that it's hard to make mistakes in future.

**Comment by nt1m** [comment]
(also one thing that would be nice to have would be content-visibility state + relevancy reasons in render tree output, potentially separately from this, to aid with debugging websites that may use content-visibility)

File: Source/WebCore/dom/Element.cpp (hunk 8 above)
Comment (rwlbuis) [reply Source/WebCore/dom/Element.cpp]: It looks like this is only used internally bu spatial navigation code. Anyway it should be safe to move this code to offsetLeft method, done.

File: Source/WebCore/dom/Element.cpp (hunk 7 above)
Comment (rwlbuis) [reply Source/WebCore/dom/Element.cpp]: Indeed it is not needed, the important thing is to make sure updateRelevancyOfContentVisibilityElements is called and style is up-to-date, fixed.

File: Source/WebCore/svg/SVGGeometryElement.cpp (hunk 6 above)
Comment (rwlbuis) [reply Source/WebCore/svg/SVGGeometryElement.cpp]: I have to ask around, since officially svg does not generate boxes, maybe not.

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:1]: These are needed to skip unneeded work in the grid algorithms and make a lot of difference in the grid perf tests.

File: Source/WebCore/rendering/RenderBlockFlow.cpp (hunk 4 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderBlockFlow.cpp]: This change is a bit vague indeed, it improves performance but I will check tomorrow if there is a more logical place that gives the same performance.

File: Source/WebCore/dom/Document.cpp (hunk 3 above)
Comment (rwlbuis) [reply Source/WebCore/dom/Document.cpp]: Yes, sounds more accurate, will have a stab at that tomorrow.

File: Source/WebCore/page/LocalFrameViewLayoutContext.h (hunk 5 above)
Comment (rwlbuis) [reply Source/WebCore/page/LocalFrameViewLayoutContext.h]: I did not see how that would work but have an idea now, will try to implement it tomorrow.

File: Source/WebCore/dom/Document.cpp (hunk 2 above)
Comment (rwlbuis) [reply Source/WebCore/dom/Document.cpp]: It is a good idea in principle I think. I implemented it, but since the Element also needs to be passed, I am not sure if it is needed in the end, since we can have a bit strange situations where ContentVisiblityForceLayout flag can be passed but not Element, or vice-versa.

However integrating the ContentVisibilityForceLayoutScope helper functionality deeper (right now it is in updateLayout) seems like a good move, ContentVisibilityForceLayoutScope is much smaller now and maybe we do not even need it in the end since it is only used in one place.

File: Source/WebCore/html/ImageInputType.cpp (hunk 9 above)
Comment (rwlbuis) [reply Source/WebCore/html/ImageInputType.cpp]: For this case I was not sure ContentVisibilityForceLayoutScope made sense in updateLayout, hence I kept it out, but looking at updateLayout again, it does make sense I think to use ContentVisibilityForceLayoutScope in updateLayout (and only there), so I was able to remove all other ContentVisibilityForceLayoutScope uses.

File: Source/WebCore/svg/SVGGeometryElement.cpp (hunk 6 above)
Comment (rwlbuis) [reply Source/WebCore/svg/SVGGeometryElement.cpp]: Playing around with this in chromium, it seems supported. I'll try to check whether this is by design, but I do think there is no proper test for this, since all tests seem to use outer svg or its container for setting content-visibility, not the svg shapes/text/groups etc.

File: Source/WebCore/rendering/RenderElement.cpp [hunk 10]
```diff
@@ -2288,5 +2288,22 @@ bool RenderElement::hasEligibleContainmentForSizeQuery() const
     return false;
 }
 
+void RenderElement::clearNeedsLayoutForDescendents()
```
Comment (nt1m) [inline Source/WebCore/rendering/RenderElement.cpp]: nit: typo

```suggestion
void RenderElement::clearNeedsLayoutForDescendants()
```

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp
Comment (nt1m) [inline Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:1]: Can we split out these changes to a new patch?

**Review by nt1m — APPROVED** [review]

File: Source/WebCore/rendering/RenderElement.cpp (hunk 10 above)
Comment (rwlbuis) [reply Source/WebCore/rendering/RenderElement.cpp]: Good eyes, fixed.

File: Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp
Comment (rwlbuis) [reply Source/WebCore/rendering/GridTrackSizingAlgorithm.cpp:1]: Sure, I filed https://bugs.webkit.org/show_bug.cgi?id=261825 to track it and will create a PR for that after this PR lands.

File: Source/WebCore/dom/Document.cpp [hunk 11]
```diff
@@ -2304,13 +2305,13 @@ bool Document::updateStyleIfNeeded()
     return true;
 }
 
-void Document::updateLayoutIgnorePendingStylesheets(OptionSet<LayoutOptions> layoutOptions)
+void Document::updateLayoutIgnorePendingStylesheets(OptionSet<LayoutOptions> layoutOptions, const Element* context)
 {
     layoutOptions.add(LayoutOptions::IgnorePendingStylesheets);
-    updateLayout(layoutOptions);
+    updateLayout(layoutOptions, layoutOptions.contains(LayoutOptions::ContentVisibilityForceLayout) ? context : nullptr);
```
Comment (nt1m) [inline Source/WebCore/dom/Document.cpp]: Passing the context unconditionally here probably doesn't hurt. It could be used for different optimizations unrelated to content visibility

Comment (rwlbuis) [reply Source/WebCore/dom/Document.cpp]: You are right, cleaner that way, adjusted.

File: Source/WebCore/rendering/RenderBlock.cpp [hunk 12]
```diff
@@ -953,6 +953,12 @@ LayoutUnit RenderBlock::marginIntrinsicLogicalWidthForChild(RenderBox& child) co
 
 void RenderBlock::layoutPositionedObject(RenderBox& r, bool relayoutChildren, bool fixedPositionObjectsOnly)
 {
+    if (isSkippedContentRoot()) {
```
Comment (nt1m) [inline Source/WebCore/rendering/RenderBlock.cpp:956]: I should have asked this earlier, but is this correct when `ContentVisibilityForceLayout` is on ? Or is this unreachable/irrelevant code when `ContentVisibilityForceLayout` is on?
